src/Backend/ReasonML.idr

```diff
@@ -0,0 +1,82 @@
+module Backend.ReasonML
+
+import Data.Vect
+import Control.Monad.State
+
+import Backend.Utils
+import Types
+import Typedefs
+
+%default partial
+%access public export
+
+formatVar : String -> String
+formatVar = (strCons '\'') . lowercase
+
+-- TODO would be nice if type was `(e : Env n) -> Vect (fst (Vect.filter isLeft e)) String`, 
+-- since it then could be used by several backends. I'm having trouble getting it to typecheck.
+getFreeVars : Env n -> String
+getFreeVars e = withSep ", " (either formatVar (const "")) $ snd $ Vect.filter isLeft e
+
+makeType : Env n -> TDef n -> String
+makeType     _ T0            = "void"
+makeType     _ T1            = "unit"
+makeType {n} e (TSum xs)     = tsum xs
+  where
+  tsum : Vect (2 + _) (TDef n) -> String
+  tsum [x, y]              = "either" ++ (parens $ (makeType e x) ++ ", " ++ makeType e y)
+  tsum (x :: y :: z :: zs) = "either" ++ (parens $ (makeType e x) ++ ", " ++ tsum (y :: z :: zs))
+makeType     e (TProd xs)    = assert_total $ parens $ withSep ", " (makeType e) xs
+makeType     e (TVar v)      = either formatVar lowercase $ Vect.index v e
+makeType     e (TMu nam _)   = let freeVars = getFreeVars e
+                                in if freeVars == "" then lowercase nam else lowercase nam ++ parens freeVars
+makeType     _ (TName nam _) = lowercase nam
+
+makeDefs : Env n -> TDef n -> State (List Name) String
+makeDefs _ T0           = pure ""
+makeDefs _ T1           = pure ""
+makeDefs e (TProd xs)   = map concat $ traverse (makeDefs e) xs
+makeDefs e (TSum xs)    = map concat $ traverse (makeDefs e) xs
+makeDefs _ (TVar v)     = pure ""
+makeDefs e (TMu nam cs) = 
+  do st <- get 
+     if List.elem nam st then pure "" 
+      else let
+         freeVars = getFreeVars e
+         dataName = if freeVars == "" then nam else nam ++ parens freeVars
+         newEnv = Right (if freeVars == "" then dataName else dataName) :: e
+         args = withSep " | " (mkArg newEnv) cs
+        in
+       do res <- map concat $ traverse {b=String} (\(_, bdy) => makeDefs newEnv bdy) cs 
+          put (nam :: st)
+          pure $ res ++ "\ntype " ++ lowercase dataName ++ " = " ++ args ++ ";\n"
+  where
+  mkArg : Env (S n) -> (Name, TDef (S n)) -> String
+  mkArg _ (cname, T1)       = cname
+  mkArg e (cname, TProd xs) = uppercase cname ++ parens (withSep ", " (makeType e) xs)
+  mkArg e (cname, ctype)    = uppercase cname ++ parens (makeType e ctype)
+makeDefs e (TName nam body) = 
+  do st <- get 
+     if List.elem nam st then pure "" 
+       else let 
+          freeVars = getFreeVars e
+          typeName = if freeVars == "" then lowercase nam else lowercase nam ++ parens freeVars
```

Comment:
Same `... ++ if ...` thing as above.

Reply:
No because `parens` gives us at least `()` back, but you're right, that's hardly desirable behavior; fixing.